File: packages/node/examples/publish.js
/*
 * A publisher. Sends a short series and exits.
 *
 *   node examples/publish.js
 *   node examples/publish.js "any message you like"
 */
import { Client } from '../src/index.js';

const client = new Client({ url: process.env.BJMSG_URL ?? 'http://127.0.0.1:8080' });

const messages = process.argv.slice(2).length
  ? process.argv.slice(2)
  : ['Hello NATS!', '...except this one is binjson over HTTP/1.1',
     { shape: 'anything binjson encodes', n: 3 }, 'goodbye'];

for (const m of messages) {
  const { index } = await client.publish('greet', m);
  console.log(`#${index}`, JSON.stringify(m));
}

/*
 * An idempotency key makes a publish safe to retry: a repeat inside the
 * broker's dedup window returns the original index rather than appending
 * a second copy.
 */
const a = await client.publish('greet', 'exactly once', { id: 'demo-key' });
const b = await client.publish('greet', 'exactly once', { id: 'demo-key' });
console.log(`same message twice -> index ${a.index} and ${b.index}`,
            `(duplicate: ${b.duplicate})`);

await client.close();

File: packages/python/examples/publish.py
"""A publisher. Sends a short series and exits.

    python examples/publish.py
    python examples/publish.py "any message you like"
"""
import os
import sys

from sukkal import Client

messages = sys.argv[1:] or [
    "Hello NATS!",
    "...except this one is binjson over HTTP/1.1",
    {"shape": "anything binjson encodes", "n": 3},
    "goodbye",
]

with Client(os.environ.get("SUKKAL_URL", "http://127.0.0.1:8080")) as client:
    for message in messages:
        result = client.publish("greet", message)
        print(f"#{result['index']} {message!r}")

    # An idempotency key makes a publish safe to retry: a repeat inside
    # the broker's dedup window returns the original index rather than
    # appending a second copy.
    a = client.publish("greet", "exactly once", id="demo-key")
    b = client.publish("greet", "exactly once", id="demo-key")
    print(f"same message twice -> index {a['index']} and {b['index']} "
          f"(duplicate: {b['duplicate']})")
